File: c/binary-search-tree/src/binary_search_tree.c
#include "binary_search_tree.h"
#include <stdlib.h>
#include <stdio.h>

void add_to_tree(node_t *current_node, int data);
node_t *initialize_node(int data);
void inorder_traversal(node_t *node, int *array, int *index);

node_t *build_tree(int *tree_data, size_t tree_data_len)
{
	if (tree_data_len == 0)
		return NULL;

	node_t *head = malloc(sizeof(node_t));
	head->data = tree_data[0];
	head->right = NULL;
	head->left = NULL;

	for (size_t i = 1; i < tree_data_len; i++) {
		add_to_tree(head, tree_data[i]);
	}

	return head;
}

void add_to_tree(node_t *current_node, int data)
{
	if (current_node->data >= data && current_node->left == NULL)
	{
		current_node->left = initialize_node(data);
		return;
	}

	if (current_node->data < data && current_node->right == NULL)
	{
		current_node->right = initialize_node(data);
		return;
	}

	add_to_tree(current_node->data >= data ? current_node->left : current_node->right, data);
}

node_t *initialize_node(int data) 
{
	node_t *node = malloc(sizeof(node_t));

	node->data = data;
	node->left = NULL;
	node->right = NULL;

	return node;
}

/*
 * Postorder traversal to free all nodes
 */
void free_tree(node_t *tree)
{
	if (tree == NULL)
		return;

	free_tree(tree->left);
	free_tree(tree->right);
	free(tree);
}

int *sorted_data(node_t *tree)
{
	int *sorted, *index;
	sorted = calloc(100, sizeof(int));
	index = calloc(1, sizeof(int));

	inorder_traversal(tree, sorted, index);

	return sorted;
}

void inorder_traversal(node_t *node, int *array, int *index) 
{
	if (node == NULL)
		return;

	inorder_traversal(node->left, array, index);
	array[(*index)++] = node->data;
	inorder_traversal(node->right, array, index);
}

File: python/binary-search-tree/binary_search_tree.py
class TreeNode:
    def __init__(self, data: str, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right

    def __str__(self) -> str:
        fmt = 'TreeNode(data={}, left={}, right={})'
        return fmt.format(self.data, self.left, self.right)


class BinarySearchTree:
    def __init__(self, tree_data: list = []):
        self.head = TreeNode(tree_data[0]) if tree_data else None

        for data in tree_data[1:]:
            self.add(self.head, TreeNode(data))

    def data(self) -> TreeNode:
        return self.head

    def sorted_data(self) -> list:
        return self.inorder_traversal(self.head, [])

    def add(self, curr_node: TreeNode, new_node: TreeNode):
        if curr_node.data >= new_node.data and curr_node.left is None:
            curr_node.left = new_node
            return

        if curr_node.data < new_node.data and curr_node.right is None:
            curr_node.right = new_node
            return

        self.add(curr_node.left if curr_node.data > new_node.data else curr_node.right, new_node)

    def inorder_traversal(self, curr_node: TreeNode, arr: list):
        if curr_node is None:
            return arr

        arr = self.inorder_traversal(curr_node.left, arr)
        arr.append(curr_node.data)
        arr = self.inorder_traversal(curr_node.right, arr)

        return arr

